

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Data Gathering &mdash; boostDM 1.0 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link href="_static/bbglab.css" rel="stylesheet" type="text/css">


  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> boostDM
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="data.html">Data Collection</a><ul>
<li class="toctree-l2"><a class="reference internal" href="data.html#tcga">TCGA</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#pcawg">PCAWG</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#cbioportal">cBioPortal</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#hartwig-medical-foundation">Hartwig Medical Foundation</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#icgc">ICGC</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#st-jude">St. Jude</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#pedcbioportal">PedcBioPortal</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#target">TARGET</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#beat-aml">Beat AML</a></li>
<li class="toctree-l2"><a class="reference internal" href="data.html#literature">Literature</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="mutational_features.html">Repository of mutational features</a><ul>
<li class="toctree-l2"><a class="reference internal" href="mutational_features.html#linear-clusters">Linear clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutational_features.html#d-clusters">3D clusters</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutational_features.html#pfam-domains">Pfam Domains</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutational_features.html#excess-of-mutations">Excess of mutations</a></li>
<li class="toctree-l2"><a class="reference internal" href="mutational_features.html#mode-of-action">Mode of action</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">boostDM</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Data Gathering</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/total.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="data-gathering">
<h1>Data Gathering<a class="headerlink" href="#data-gathering" title="Permalink to this headline">¶</a></h1>
<div class="section" id="tcga">
<h2>TCGA<a class="headerlink" href="#tcga" title="Permalink to this headline">¶</a></h2>
<p>TCGA somatic mutations (mc3.v0.2.8 version) were downloaded from
(<a class="reference external" href="https://gdc.cancer.gov/about-data/publications/pancanatlas">https://gdc.cancer.gov/about-data/publications/pancanatlas</a>). We
then grouped mutations according to their patient’s cancer type into 32
different cohorts. Additionally, we filtered out somatic mutations that
did not pass the somatic filtering from TCGA (i.e., column FILTER !=
“PASS”).</p>
</div>
<div class="section" id="pcawg">
<h2>PCAWG<a class="headerlink" href="#pcawg" title="Permalink to this headline">¶</a></h2>
<p>PCAWG somatic mutations were downloaded from the International Cancer
Genome Consortium (ICGC) data portal
(<a class="reference external" href="https://dcc.icgc.org/releases/PCAWG/consensus_snv_indel/">https://dcc.icgc.org/releases/PCAWG/consensus_snv_indel/</a>). Note
that only ICGC samples can be freely downloaded from this site, the TCGA
portion of the callsets is controlled data. Instructions on how to
obtain them can be found in the same webpage.</p>
</div>
<div class="section" id="cbioportal">
<h2>cBioPortal<a class="headerlink" href="#cbioportal" title="Permalink to this headline">¶</a></h2>
<p>Somatic mutations from Whole Exome Sequencing (WXS) and Whole Genome
Sequencing (WGS) cohorts uploaded in cBioPortal that were not part of
any other projects included in the analysis (i.e., TCGA, PCAWG, ST. JUDE
or HARTWIG) were downloaded (download date 2018/09/01;
<a class="reference external" href="http://www.cbioportal.org/datasets">http://www.cbioportal.org/datasets</a>). We then created cohorts following
the next criteria:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>Cohorts with a limited number of samples (i.e., lower than 30</dt><dd><p>samples) associated to cancer types with extensive representation
(such as Breast cancer, Prostate cancer, Colorectal
adenocarcinoma…) across the compendium of cohorts were removed.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Samples were uniquely mapped into one single cohort. If the same</dt><dd><p>sample was originally included in two cohorts, we removed the
sample from one of them.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Samples not sequenced from human cancer biopsies were discarded (cell</dt><dd><p>lines, xenografts, normal tissue, etc.)</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When patient information was available, only one sample from the same</dt><dd><p>patient was selected. The criteria to prioritize samples from the
same patient was: WXS over WGS; untreated over treated, primary
over metastasis or relapse and, finally, by alphabetical order.
When there is no patient information we assume that all patients
have only one sample in the cohort.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When sequencing platform information was available, samples from the</dt><dd><p>same study but with different sequencing platforms were further
subclassified into WXS and WGS datasets (only if the resulting
cohorts fulfilled the requirements herein described; otherwise,
the samples were discarded).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When variant calling information was available, samples from the same</dt><dd><p>cohort and sequencing type were further classified according to
their calling algorithm (VarScan, MuTect, etc.). If the resulting
cohorts for each subclass fulfilled the requirements herein
described, the samples were included;otherwise, the samples were
discarded. When variant calling information was not available we
assumed that all the samples went through the same pipeline.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When treatment information was available, samples from the same</dt><dd><p>cohort, sequencing type, calling algorithm were further classified
according to their treatment status (i.e, treated versus
untreated). If the resulting cohorts from the subclassification
fulfilled the requirements herein described, the samples were
included;otherwise, the samples were discarded. When information
was not available we assumed that samples had not been treated.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When biopsy information was available, samples from the same cohort,</dt><dd><p>sequencing type, calling algorithm, treatment status were further
classified according to their biopsy type (i.e, primary, relapse
or metastasis). If the resulting datasets from the
subclassification fulfilled the requirements herein described, the
samples were included; otherwise, the samples were discarded. When
information was not available we assumed that the biopsy type of
the sample was primary.</p>
</dd>
</dl>
</li>
</ol>
</div>
<div class="section" id="hartwig">
<h2>HARTWIG<a class="headerlink" href="#hartwig" title="Permalink to this headline">¶</a></h2>
<p>Somatic mutations of metastatic WGS from Hartwig Medical Foundation were
downloaded through their platform (download date 2019/05/02). Datasets
were split according to their primary site. Samples from unknown primary
sites (i.e., None, Nan, Unknown, Cup, Na), double primary or aggregating
into cohorts of fewer than 5 samples (i.e., Adrenal, Myeloid, Thymus and
Eye) were not considered. A total of 25 different cohorts were created.</p>
</div>
<div class="section" id="icgc">
<h2>ICGC<a class="headerlink" href="#icgc" title="Permalink to this headline">¶</a></h2>
<p>Somatic mutations from Whole Exome Sequencing (WXS) and Whole Genome
Sequencing (WGS) studies uploaded in ICGC Data Portal
(<a class="reference external" href="https://dcc.icgc.org/repositories">https://dcc.icgc.org/repositories</a>) not overlapping with other
projects included in the analysis (i.e., TCGA, PCAWG, CBIOP or ST. JUDE)
were downloaded on 2018/01/09. We then created cohorts following the
criteria used for the cBioPortal datasets (cBioPortal).</p>
</div>
<div class="section" id="st-jude">
<h2>ST. JUDE<a class="headerlink" href="#st-jude" title="Permalink to this headline">¶</a></h2>
<p>Somatic mutations from Whole Exome Sequencing (WXS) and Whole Genome
Sequencing (WGS) of Pediatric Cancer Genome Project uploaded in the St.
Jude Cloud (<a class="reference external" href="https://www.stjude.cloud/data.html">https://www.stjude.cloud/data.html</a>) were downloaded on
2018/07/16. Cohorts were created according to their primary site and
their biopsy type (i.e., primary, metastasis and relapse). Resulting
datasets with fewer than 5 samples were discarded.</p>
</div>
<div class="section" id="pedcbioportal">
<h2>PedcBioPortal<a class="headerlink" href="#pedcbioportal" title="Permalink to this headline">¶</a></h2>
<p>Somatic mutations from Whole Exome Sequencing (WXS) and Whole Genome
Sequencing (WGS) studies uploaded in PedcBioPortal that were not part of
any other projects included in the analysis (i.e., ST. JUDE or CBIOP)
were downloaded (download date
2018/10/01;<a class="reference external" href="http://www.pedcbioportal.org/datasets">http://www.pedcbioportal.org/datasets</a>). We then created
cohorts following the criteria described in the cBioPortal dataset
(cBioPortal).</p>
</div>
<div class="section" id="target">
<h2>TARGET<a class="headerlink" href="#target" title="Permalink to this headline">¶</a></h2>
<p>Somatic SNVs from WXS and WGS of two TARGET studies, Neuroblastoma (NB)
and Wilms Tumor (WT), from the TARGET consortium were downloaded from
the GDC Data Portal (download date 2019/03/07).</p>
</div>
<div class="section" id="beat-aml">
<h2>Beat AML<a class="headerlink" href="#beat-aml" title="Permalink to this headline">¶</a></h2>
<p>We downloaded unfiltered somatic mutations from samples included in the
Beat AML study. We next applied the following criteria to create our
Beat AML cohort:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>We focused on somatic single nucleotide variants from VarScan2 using</dt><dd><p>skin as normal control. All samples that did not belong to this
class were not further analyzed.</p>
</dd>
</dl>
</li>
<li><p>Samples from relapses were filtered out.</p></li>
<li><p>Samples from bone-marrow transplants were discarded.</p></li>
<li><dl class="simple">
<dt>If there were several samples per patient fulfilling the points 1-3,</dt><dd><p>we selected the first in</p>
</dd>
</dl>
</li>
</ol>
<blockquote>
<div><p>chronological order.</p>
</div></blockquote>
<p>257 independent samples of Beat AML tumors composed our Beat AML cohort.</p>
</div>
<div class="section" id="literature">
<h2>LITERATURE<a class="headerlink" href="#literature" title="Permalink to this headline">¶</a></h2>
<p>We also manually collected publicly available cohorts from the
literature. Each cohort was filtered following the same steps than
mentioned above for the cBioPortal dataset.</p>
<p>For a summary of all datasets used in the current release of intogen
download Table XX.</p>
</div>
</div>
<div class="section" id="preprocessing">
<span id="section-1"></span><h1>Preprocessing<a class="headerlink" href="#preprocessing" title="Permalink to this headline">¶</a></h1>
<p>Given the heterogeneity of the datasets analyzed in the current release
of intOGen (resulting from e.g. differences in the genome aligners,
variant calling algorithms, sequencing coverage, sequencing strategy),
we implemented a robust pre-processing strategy aiming at reducing
possible biases. Specifically, we conducted the following filtering
steps:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>The pipeline is configured to run using GRCh38 as reference genome.</dt><dd><p>Therefore, for each input dataset the pipeline requires that the
reference genome is defined. Datasets using GRCh37 as reference
genome were lifted over using PyLiftover
(<a class="reference external" href="https://pypi.org/project/pyliftover/">https://pypi.org/project/pyliftover</a>/;
version 0.3) to GRCh38. Mutations failing to liftover from GRCh37
to GRCh38 were discarded.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We removed mutations with equal alternate and reference alleles,</dt><dd><p>duplicated mutations within the sample sample, mutations with ‘N’
as reference or alternative allele, mutations with a reference
allele not matching its reference genome and mutations within
non-canonical chromosomes (i.e., mutations outside chr1 to chr22,
chrX and chrY).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Additionally, we removed mutations with low pileup mappability, i.e.</dt><dd><p>mutations in regions that could potentially map elsewhere in the
genome. For each position of the genome we computed the pileup
mappability, defined as the average uniqueness of all the possible
reads of 100bp overlapping a position and allowing up to 2
mismatches. This value is equal to 1 if all the reads overlapping
a mutation are uniquely mappable while it is close to 0 if most
mapping reads can map elsewhere in the genome. Positions with a
pileup mappability lower than 0.9 were removed from further
analyses.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We filtered out multiple samples from the same donor. The analysis of</dt><dd><p>positive selection in tumors requires that each sample in a cohort
is independent from the other samples. That implies that if the
input dataset includes multiple samples from the same patient
–resulting from different biopsy sites, time points or sequencing
strategies– the pipeline automatically selects the first
according to its alphabetical order. Therefore, all mutations in
the discarded samples are not considered anymore.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We also filtered out hypermutated samples. Samples carrying more than</dt><dd><p>1000 mutations for WXS and 10000 for WGS and a mutation count
greater than 1.5 times the interquartile range length above the
third quartile in their respective dataset were considered
hypermutated and therefore removed from further analyses.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Datasets with filtered synonymous variants are not runnable. Most</dt><dd><p>cancer driver identification methods need synonymous variants to
fit a background mutation model. Therefore, datasets with less
than 5 synonymous and datasets with a missense/synonymous ratio
greater than 10 were excluded .</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When the Variant Effect Predictor (VEP) mapped one mutation into</dt><dd><p>multiple transcripts associated with HUGO symbols, we selected the
canonical transcript of the first HUGO symbol in alphabetical
order.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We also discarded mutations mapping into genes without canonical</dt><dd><p>transcript in VEP.92.</p>
</dd>
</dl>
</li>
</ol>
</div>
<div class="section" id="methods-for-cancer-driver-gene-identification">
<h1>Methods for cancer driver gene identification<a class="headerlink" href="#methods-for-cancer-driver-gene-identification" title="Permalink to this headline">¶</a></h1>
<p>The current version of the intOGen pipeline uses six cancer driver
identification methods (hereinafter DIMs) to identify cancer driver
genes from somatic point
mutations:<a class="reference external" href="https://github.com/im3sanger/dndscv">dNdScv</a>and<a class="reference external" href="http://genetics.bwh.harvard.edu/cbase/index.html">CBaSE</a>,
which test for mutation count bias in genes against a background
mutation rate model that incorporates regional genomic covariates, the
tumour mutational process and mutation consequence
types;<a class="reference external" href="http://bbglab.irbbarcelona.org/oncodriveclustl/home">OncodriveCLUSTL</a>,
which tests for significant clustering of mutations in the protein
sequence; smRegions, which tests for enrichment of mutations in protein
functional domains; HotMAPS, which tests for significant clustering of
mutations in the 3D protein structure;
and<a class="reference external" href="http://bbglab.irbbarcelona.org/oncodrivefml/home">OncodriveFML</a>,
which tests for functional impact bias of the observed mutations. Next
we briefly describe the rationale and the configuration used to run each
DIM.</p>
<p><strong>dNdScv</strong></p>
<p>dNdScv [REF] asserts gene-specific positive selection by inferring the
ratio of non-synonymous to synonymous substitutions (dN/dS, hereinafter
termed <span class="math notranslate nohighlight">\(\omega\)</span>) in the coding region of each gene. The method
resorts to a Poisson-based hierarchical count model that can correct
for: i) the mutational processes operative in the tumor determined by
the mutational profile of single-nucleotide substitutions with its
flanking nucleotides, ii) the distribution of consequence types at the
CDS per gene, and iii) the regional variability of the background
mutation rate; it incorporates information about 10 histone marks from
69 cell lines obtained in ENCODE project [REF].</p>
<p>We downloaded (release date 2018/10/12) and built a new reference
database based on the list canonical transcripts defined by VEP.92
(GRCh38). We then used this reference database to run dNdScv on all
datasets of somatic mutations using the default setting of the method.</p>
<p><strong>OncodriveFML</strong></p>
<p>OncodriveFML [ref] is a tool that aims to detect genes under positive
selection by analysing the functional impact bias of the observed
somatic mutations. Briefly, OncodriveFML consists of three steps: in the
first step, it computes the average Functional Impact (FI) score (in our
pipeline we used CADD v1.4) of coding somatic mutations observed in gene
of interest across a cohort of tumor samples. In the next step, sets of
mutations of the same size as the number of mutations observed in the
gene of interest are randomly sampled following the tri-nucleotide
probabilities. This sampling is repeated N times (N=106 in our
configuration) to generate expected average scores across all genes of
interest. Finally, it compares the observed average FI score with the
expected from the simulations in the form of an empirical p-value. The
p-values are then adjusted with a multiple testing correction using the
Benjamini–Hochberg (FDR).</p>
<p>OncodriveCLUSTL</p>
<p>OncodriveCLUSTL is a sequence-based clustering algorithm to detect
significant linear clustering bias of the observed somatic mutations
[REF]. Briefly, OncodriveCLUSTL first maps somatic single nucleotide
variants observed in a cohort to the genomic element under study. After
smoothing mutations along its genomic sequence using a Tukey kernel
based density function, clusters are detected and scored taking into
account the number and distribution of mutations observed. A score for
each genomic element is obtained by adding up the scores of its
clusters. To estimate the significance of the observed clustering
signals, mutations are locally randomized using tri- or penta-nucleotide
context probabilities calculated from the input cohort.</p>
<p>For this analysis, OncodriveCLUSTL version 1.1.1 was run for the set of
defined canonical transcripts bearing 2 or more SNVs mapping the
mutations file as follows: smoothing, clustering windows were kept as
default (11bp). The different consecutive coding sequences contained on
each transcript were concatenated to allow the algorithm to detect
clusters of 2 or more SNVs expanding two exons in a transcript.
Simulations were carried out using previously computed mutational
profiles (see signatures above). All cohorts were run using
tri-nucleotide context SNVs profiles except for cutaneous melanomas,
where penta-nucleotide profiles were calculated. Default randomization
windows of 31bp length were not allowed to expand beyond the coding
sequence boundaries (e.g., windows overlapping part of an exon and an
intron were shifted to fit inside the exon). A total number of 1,000
simulations per transcript were performed. Transcripts with q-value &lt;
0.01 were considered significant.</p>
<p><strong>CBaSe</strong></p>
<p>CBaSe [REF] asserts gene-specific positive and negative selection by
modelling the distribution of non-synonymous mutation categories under
neutral selection resorting to a Poisson-based hierarchical modelling
approach. As in the case of dNdScv, the method also allows for
correction by i) the mutational processes operative in the tumor as
defined by the mutational profile of single-nucleotide substitutions
–with either tri- or penta- nucleotide context–, ii) the distribution
of consequence types per gene, and iii) regional variability of the
neutral mutation rate. The method theoretically corrects for both known
and unknown covariates of the regional mutation rate, but in practice
the method is sensitive to the count of synonymous mutations. Finally,
the method allows 6 different models based on distinct prior
alternatives for the distribution of the regional mutation rate.</p>
<p>We run a modified version of the CBaSe script to fit the specific needs
of our pipeline. The main modification was adding a clause to
automatically select a regional mutation rate prior distribution that
suits the size of the dataset. Based on the total mutations count in the
dataset, the method runs either an inverse-gamma (mutation count &lt;
12,000), an exponential-inverse-gamma mixture (12,000 &lt; mutation count &lt;
65,000) or a gamma-inverse-gamma mixture (mutation count &gt; 65,000) as
regional mutation rate priors (following communication by Donate
Weghorn, CBaSe’s first author). Furthermore, we also skip the negative
selection analysis and modified the output formatting, including a new
Benjamini-Hochberg FDR column with q-values.</p>
<p>HotMaps3D</p>
<p>HotMAPS algorithm (HotMAPS-1.1.3 version REF) was modified to include a
new background model that more accurately represents the probability of
somatic mutations in a particular cancer type. The original HotMAPS
algorithm, assumes that all amino-acid substitutions in a protein
structure are equally probable. Herein, we implemented a modified
version of the algorithm where the mutation probability depends on the
trinucleotide context and the mutational processes that are operating in
that cohort of samples. Briefly, for each analyzed protein structure
harbouring missense mutations, the same number of simulated mutations
were randomly generated within the protein structure considering the
precomputed tri-nucleotide frequencies in that cohort. This
pseudo-random sampling was performed N times (N=100,000 in our
configuration) deriving into the background model to compare with the
observed mutational data. The rest of HotMAPS algorithm was kept as it
was.</p>
<p>We downloaded the pre-computed mapping of GRCh37 coordinates into
structure residues from the Protein Data Bank (PDB)
(<a class="reference external" href="http://karchinlab.org/data/HotMAPS/mupit_modbase.sql.gz">http://karchinlab.org/data/HotMAPS/mupit_modbase.sql.gz</a>). We also
downloaded all protein structures from the PDB (download date
2019/09/20) alongside all human protein 3D models from Modeller
(download date 2019/09/20;
<a class="reference external" href="ftp://salilab.org/databases/modbase/projects/genomes/H_sapiens/2013/H_sapiens_2013.tar.xz">ftp://salilab.org/databases/modbase/projects/genomes/H_sapiens/2013/H_sapiens_2013.tar.xz</a>
and
<a class="reference external" href="ftp://salilab.org/databases/modbase/projects/genomes/H_sapiens/2013/ModBase_H_sapiens_2013_refseq.tar.xz">ftp://salilab.org/databases/modbase/projects/genomes/H_sapiens/2013/ModBase_H_sapiens_2013_refseq.tar.xz</a>).
We then annotated the structures following the steps described in
HotMAPS tutorial
(<a class="reference external" href="https://github.com/KarchinLab/HotMAPS/wiki/Tutorial-(Exome-scale">https://github.com/KarchinLab/HotMAPS/wiki/Tutorial-(Exome-scale</a>)).</p>
<p>Since HotMAPS configuration files are pre-built in GRCh37 coordinates
and our pipeline is designed to run using GRCh38, for each input cohort,
we first converted input somatic mutations to GRCh37, executed the
HotMAPS algorithm and transformed the output coordinates to GRCh38. All
conversions are done using PyLiftover tool.</p>
<p>smRegions</p>
<p>smRegions is a method developed to detect linear enrichment of somatic
mutations in user-defined regions of interest [ref]. Briefly, smRegions
first counts the number of nonsynonymous mutations overlapping with a
Pfam domain in a particular protein. Next, these nonsynonymous variants
are randomly re-sampled N times (N=1,000 in our configuration) along the
nucleotide sequence of the gene, following the probability of mutation
of each base, derived from the pre-computed tri-nucleotide frequencies.
The observed and average number of simulated mutations in the Pfam
domain and outside of it are compared using a G-test of goodness-of-fit,
from which the smRegions p-value is derived. We discarded those domains
with a number of observed mutations lower than the average from the
randomizations. The p-values were adjusted with a multiple testing
correction using the Benjamini–Hochberg procedure. Therefore, we
confined the analysis to Pfam domains with a number of observed
mutations higher or equal than the mean simulated number of mutations in
the re-sampling.</p>
<p>To create the database of genomic coordinates of Pfam domains we
followed the next steps: i) we gathered the first and last amino acid
positions of all Pfam domains for canonical transcripts (VEP.92) from
BioMart; ii) for each Pfam domain we mapped the first and last amino
acid positions into genomic coordinates using TransVar –using GRCh38 as
reference genome–; iii) we discarded Pfam domains failing to map either
the first or last amino acid positions into genomic coordinates.</p>
<p>smRegions was conceptually inspired by e-driver [ref], although
significant enhancements were introduced. Particularly, i) our
background model accounts for the observed tri-nucleotide frequencies
rather than assuming that all mutations are equally likely; ii) the
statistical test is more conservative; iii) Pfam domains are part of the
required input and can be easily updated by downloading the last Pfam
release iv) the method can be configured to any other setting that aims
to detect genes possibility selected by enrichment of mutations in
pre-defined gene regions.</p>
</div>
<div class="section" id="combining-the-outputs-of-driver-identification-methods">
<h1>Combining the outputs of driver identification methods<a class="headerlink" href="#combining-the-outputs-of-driver-identification-methods" title="Permalink to this headline">¶</a></h1>
<p>Our goal is to provide a catalogue of driver elements which
appropriately reflects the consensus from the DIMs we run.</p>
<p>To combine the results of individual statistical tests, p-value
combination methods continue to be a standard approach in the field:
e.g., Fisher [REF], Brown [REF] and Stouffer Z-score [REF] methods have
been used for this purpose. These methods are useful for combining
probabilities in meta-analysis, hence to provide a ranking based on
combined significance under statistical grounds. However, the
application of these methods may bear some caveats:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>The ranking resulting from p-value combination may lead to</dt><dd><p>inconsistencies when compared to the individual rankings, i.e.,
they may yield a consensus ranking that does not preserve
recurrent precedence relationships found in the individual
rankings.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Some methods, like Fisher’s or Brown’s method, tend to bear</dt><dd><p>anti-conservative performance, thus leading to many likely false
discoveries.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Balanced (non-weighted) p-value combination methods may lead to</dt><dd><p>faulty results just because of the influence of one or more DIM
performing poorly for a given dataset.</p>
</dd>
</dl>
</li>
</ol>
<p>Weighted methods to combine p-values, like the weighted Stouffer
Z-score, provide some extra room for proper balancing, in the sense of
modelling the relative credibility of each DIM. We reasoned that any
good operational criteria to allocate weights should satisfy the
following requirements: i) provide a specific weighting for each cohort,
thereby allowing the relative credibility of a DIM to depend on the
cohort; ii) reflect prior knowledge about known bona-fide driver genes;
iii) reflect prior knowledge about the criteria that each DIM employed
to yield its output.</p>
<p>Our approach works independently for each cohort: to create a consensus
list of driver genes for each cohort, we first determine how credible
each DIM is when applied to this specific cohort, on the basis of how
many bona fide cancer genes reported in the COSMIC Cancer Gene Census
database (CGC) are highly ranked according to the DIM. Once the
credibility has been quantified, we use a weighted method for combining
the p-values that each DIM produces for each candidate gene. This
combination takes the DIMs credibility into account. Based on the
combined p-values, we conduct FDR correction to conclude a ranking of
candidate driver genes alongside q-values.</p>
<p>The relative credibility for each method is based on the ability of the
method to give precedence to well-known genes already collected in the
CGC catalogue of driver genes. As each method yields a ranking of driver
genes, these lists can be combined using a voting system –based on the
so-called Schulze’s method. The method allows us to consider each method
as a voter with some voting rights (weighting) which casts ballots
containing a list of candidates sorted by precedence. Schulze’s method
takes precedence information from each individual method and produces a
new consensus ranking [ref].</p>
<p>Instead of conducting balanced voting, we tune the voting rights of the
methods so that we maximize the enrichment of CGC genes at the top
positions of the consensus list upon voting. In order to prevent
degenerate solutions, we impose some constraints so that every method
contributes with a minimum share. We also limit the combined share of
coalitions of methods based on similar signals of positive selection.
The solution voting rights are deemed the relative credibility for each
method.</p>
<p>Upon selection of a catalogue of bona-fide known driver elements (CGC
catalogue of driver genes) we can provide a score for each ranking
<span class="math notranslate nohighlight">\(R\)</span> of genes as follows:</p>
<p><span class="math notranslate nohighlight">\(E(R)\  = \frac{p_{i}}{\log(i + 1)}\)</span></p>
<p>where <span class="math notranslate nohighlight">\(p_{i}\)</span>is the proportion of elements with rank higher than
<span class="math notranslate nohighlight">\(i\)</span> which belong to CGC and N is a suitable threshold to consider
only the N top ranked elements. Using <span class="math notranslate nohighlight">\(E\)</span> we can define a function
<span class="math notranslate nohighlight">\(f\)</span> that maps each weighting vector <span class="math notranslate nohighlight">\(w\)</span> (in the 5-simplex)
to a value <span class="math notranslate nohighlight">\(E(R_{w})\)</span> where <span class="math notranslate nohighlight">\(R_{w}\)</span> denotes the consensus
ranking obtained by applying Schulze’s voting with voting rights given
by the weighting vector w.</p>
<p>Finally we are bound to find a good candidate for
<span class="math notranslate nohighlight">\(\widehat{w_{}} = argmax_{}(f)\)</span>. In order to prevent spurious
outcomes that could be explained by the concordant detection of driver
signals by methods with similar underlying assumptions, we impose a set
of mild ad-hoc constraints. Specifically, we set the following
constraints: i) all relative weights <span class="math notranslate nohighlight">\(\geq\)</span> 0.05; ii) w(dNdScv) +
w(cBaSe) <span class="math notranslate nohighlight">\(\leq\)</span> 0.5; iii) w(OncodriveCLUSTL) + w(HotMaps3D) +
w(smRegions) <span class="math notranslate nohighlight">\(\leq\)</span> 0.5; w(OncodriveFML) <span class="math notranslate nohighlight">\(\leq\)</span> 0.5.</p>
<p>Optimization is then carried out in two steps. The first step finds a
candidate <span class="math notranslate nohighlight">\(\widehat{w_{0}}\)</span> by exhaustive search in a rectangular
grid satisfying the constraints defined above (grid step=0.05). In the
second step we take <span class="math notranslate nohighlight">\(\widehat{w_{0}}\)</span> as the seed for a stochastic
hill-climbing procedure (using Python’s scipy.optimize “basinhopping”,
method=SLSQP and stepsize=0.05).</p>
<p>Using the relative weight estimate that yields a maximum value of the
objective function f we combined the p-values resorting to the weighted
Stouffer Z-score method. Thereafter we performed Benjamini-Hochberg FDR
correction with the resulting combined p-values, yielding one q-value
for each genomic element. If the element belongs to CGC, we computed its
q-value using only the collection of p-values computed for CGC genes
(hereto referred to as CGC q-value). Otherwise, we computed the q-value
using all the computed p-values (hereto referred simply to as q-value).</p>
<p>To finalize the analysis we considered only genes with at least two
mutated samples in the cohort under analysis. These genes were
classified into four groups according to the level of evidence in that
cohort that the gene harbours positive selection.</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>The first group, named as TIER1, contained genes showing high</dt><dd><p>confidence and agreement in their positive selection signals.
Given the ranked list of genes obtained by the Schulze voting,
TIER1 comprises all the ranked genes whose ranking is higher than
the first gene with combined q-value lower than a specific
threshold (by default threshold=0.05). The second group, name as
TIER2, was devised to contain known cancer driver genes, showing
mild signals of positive selection, that were not included in
TIER1. More in detail, we defined TIER2 genes as those CGC genes,
not included in TIER2, whose CGC q-value was lower than a given
threshold (default CGC q-value=0.25). CGC q-value is computed by
performing multiple test correction of combined p-values
restricted to CGC genes. The third group, are genes not included
in TIER1 or TIER2 with scattered signals of positive selection,
frequently coming from one single method. Particularly, given the
ranked list of genes by the Schulze voting, TIER3 was composed of
all the ranked genes with q-value lower than a given threshold (by
default threshold=0.05) whose ranking is higher than TIER1 last
gene position and lower than the rejection ranking position. The
rejection ranking position is defined as the ranking position for
which all elements have a q-value lower than the input threshold
(by default threshold=0.05). Finally, other genes not included in
the aforementioned classes are considered non-driver genes.</p>
</dd>
</dl>
</li>
</ol>
<p><a class="reference internal" href="media/image1.png"><img alt="image1" src="media/image1.png" style="width: 6.5in; height: 1.83333in;" /></a></p>
</div>
<div class="section" id="combination-evaluation">
<h1>Combination evaluation<a class="headerlink" href="#combination-evaluation" title="Permalink to this headline">¶</a></h1>
<p>We carried out a benchmark of intOGen’s combination against other
procedures in terms of two scoring systems: enrichment of top-ranked
genes in bona-fide known cancer genes (CGC-Score); enrichment in known
false discovery artifacts (Negative-Score), e.g. long, late-replicating,
structural and/or inactive genes.</p>
<p>We carried out the evaluation across the 32 Whole-Exome cohorts of the
TCGA PanCanAtlas project (downloaded
from<a class="reference external" href="https://gdc.cancer.gov/about-data/publications/pancanatlas">https://gdc.cancer.gov/about-data/publications/pancanatlas</a>). By
using these cohorts we could ensure that sequence coverage, sequence
alignment and somatic mutation calling was performed using the same
methodology.</p>
<p>The Cancer Genes Census –version v87– was downloaded from the COSMIC
data portal (<a class="reference external" href="https://cancer.sanger.ac.uk/census">https://cancer.sanger.ac.uk/census</a>) and used as a
positive set of known cancer driver genes.</p>
<p>We created a catalog of genes that are known not to be involved in
cancerogenesis. This set includes very long genes (HMCN1, TTN, OBSCN,
GPR98, RYR2 and RYR3), and a list of olfactory receptors from Human
Olfactory Receptor Data Exploratorium (HORDE)
(<a class="reference external" href="https://genome.weizmann.ac.il/horde/">https://genome.weizmann.ac.il/horde/</a>). In addition, for all TCGA
cohorts, we added non-expressed genes, defined as genes where at least
80% of the samples showed a negative log2 RSEM. Expression data for TCGA
was downloaded
from<a class="reference external" href="https://gdc.cancer.gov/about-data/publications/pancanatlas">https://gdc.cancer.gov/about-data/publications/pancanatlas</a>.</p>
<p>We defined a score, so-called CGC-Score, that is intended to measure the
enrichment of CGC elements in the top positions of the ranking;
specifically given a ranking <span class="math notranslate nohighlight">\(R\)</span> mapping each element to a rank,
we define the CGC-Score of <span class="math notranslate nohighlight">\(R\)</span> as</p>
<p><span class="math notranslate nohighlight">\(\text{CGC}\text{score}(R)\  = \frac{p_{i}}{log(i + 1)}\)</span> /
:math:<a href="#id1"><span class="problematic" id="id2">``</span></a><span class="math notranslate nohighlight">\(\frac{1}{log(i + 1)}\)</span></p>
<p>where <span class="math notranslate nohighlight">\(p_{i}\)</span>is the proportion of elements with rank
<span class="math notranslate nohighlight">\(\leq i\)</span> that belong to CGC and <span class="math notranslate nohighlight">\(N\)</span>is a suitable threshold
to consider just the top elements in the ranking (by default N=40).</p>
<p>We estimated the CGC-Score across TCGA cohorts for the individual
methods ranking and the combined ranking.</p>
<p>Similarly, we defined another score, so-called Negative-Score, that aims
to measure the proportion of non-cancer genes among the top positions in
the ranking. Particularly, given a ranking <span class="math notranslate nohighlight">\(R\)</span> mapping each
element to a rank, we define the Negative-Score of <span class="math notranslate nohighlight">\(R\)</span>as</p>
<p><span class="math notranslate nohighlight">\(\text{CGC}\text{score}(R)\  = \frac{p_{i}}{log(i + 1)}\)</span> /
:math:<a href="#id3"><span class="problematic" id="id4">``</span></a><span class="math notranslate nohighlight">\(\frac{1}{log(i + 1)}\)</span></p>
<p>where <span class="math notranslate nohighlight">\(p_{i}\)</span>is the proportion of elements with rank
<span class="math notranslate nohighlight">\(\leq i\)</span> that belong to the negative set and <span class="math notranslate nohighlight">\(N\)</span> is a
suitable threshold to consider just the top elements in the ranking (by
default N=40). We estimated the Negative-Score across TCGA cohorts for
the individual methods ranking and the combined ranking.</p>
<p>When compared to individual methods, the full combination achieves
highest enrichment of CGC genes among top ranked genes. When compared to
commonly used alternative combination methods based on the same
individual outcomes, our proposed combination tended to yield highest
CGC-Score. We also checked the enrichment of known false discovery
artifacts: we found that our combination method tended to yield lower
Negative-Score than individual methods.</p>
<p>We aimed to estimate the contribution of each method’s ranking to the
final consensus ranking. To address this question, we measured the
effect of removing one DIM at a time from the combination by computing
the ratio between the CGC-Score of the combination excluding the DIM and
the GCG-Score of the full combination. This was done for all DIMs across
TCGA cohorts. Only those DIMs that positively contribute to the combined
ranking attained by intOGen show a ratio below one.</p>
<p>We also assessed the effect on the CGC-Score upon leaving each method
out before combining, one method at a time: with few exceptions, the
effect of leaving any method out was detrimental to the CGC-Score,
meaning that in general all the methods are effectively required to
reach the optimal consensus attained. Also, there is substantial
variability across TCGA cohorts in terms of DIM contributions to the
full combination.</p>
<p>In summary, all methods contributed positively to the combinatorial
performance across TCGA supporting our methodological choice for the
individual driver discovery methods.</p>
<p>We compared the CGC-Score and Negative-Score of our combinatorial
selection strategy against other methods frequently used employed to
produce ranking combinations, either based on ranking information –such
as Borda Count (REF)– or based on statistical information –such as
Fisher (REF) or Brown (REF, REF) methods. Hereto, we briefly describe
the rationale of the four methods we used to benchmark our ranking. For
the sake of compact notation, let’s denote the rank and p-value of gene
<span class="math notranslate nohighlight">\(g\)</span> produced by method <span class="math notranslate nohighlight">\(m_{i}\)</span> as <span class="math notranslate nohighlight">\(r_{i,\ g}\)</span> and
<span class="math notranslate nohighlight">\(p_{i,\ g}\)</span>, respectively.</p>
<p>Borda Count: For each ranked item <span class="math notranslate nohighlight">\(g\)</span> and method <span class="math notranslate nohighlight">\(m_{i},\)</span> it
assigns a score <span class="math notranslate nohighlight">\(s_{i,\ g} = N - l_{i,\ g},\)</span> where <span class="math notranslate nohighlight">\(N\)</span>
stands for the total number of items to rank and <span class="math notranslate nohighlight">\(l_{i,\ g}\)</span> is
the number of items ranked below <span class="math notranslate nohighlight">\(g\)</span> according to method
<span class="math notranslate nohighlight">\(m_{i}\)</span>. For each item <span class="math notranslate nohighlight">\(g\)</span> an overall score
<span class="math notranslate nohighlight">\(s_{g}{\  = \ s}_{1,\ g} + \ldots + s_{k,\ g}\)</span> can then be
computed for each <span class="math notranslate nohighlight">\(g,\)</span> whence a ranking is obtained by descending
sort.</p>
<p>Fisher: It is based on the p-values <span class="math notranslate nohighlight">\(p_{i,\ g}\)</span>. For each item
<span class="math notranslate nohighlight">\(g\)</span> the method produces a new combined p-value by computing the
statistic:</p>
<p><span class="math notranslate nohighlight">\(F_{g} = - 2\log\ p_{i,\ g}\  \sim \ \chi_{2k}^{2}\)</span>.</p>
<p>Under the null hypothesis, <span class="math notranslate nohighlight">\(F_{g}\)</span> are distributed as a chi-square
with <span class="math notranslate nohighlight">\(2k\)</span> degrees of freedom, whence a p-value, which in turn
yields a raking by ascending sort. Its applicability is limited by the
assumption that the methods provide independent significance tests.</p>
<p>Brown: This method overcomes the independence requirement of Fisher’s
method by modeling the dependencies between the statistical tests
produced by each method, specifically by estimating the covariance
<span class="math notranslate nohighlight">\(\Omega_{i,\ j} = cov( - 2\log\text{\ p}_{i,\ g},\  - 2\log\ p_{j,\ g}).\)</span>
Brown’s method (REF) and its most recent adaptation (REF) have been
proposed as less biased alternatives to Fisher.</p>
<p>We then computed the CGC-Score and Negative-Score based on the consensus
ranking from the aforementioned combinatorial methods and compared them
to our Schulze’s weighted combination ranking across all TCGA cohorts.
Our combinatorial approach met a larger enrichment in known cancer genes
for 29/32 (90%) TCGA cohorts [Figure].</p>
</div>
<div class="section" id="drivers-postprocessing">
<h1>Drivers postprocessing<a class="headerlink" href="#drivers-postprocessing" title="Permalink to this headline">¶</a></h1>
<p>The intOGen pipeline outputs a ranked list of driver genes for each
input cohort. We aimed to create a comprehensive catalog of driver genes
per tumor type from all the cohorts included in this version.</p>
<p>Then, we performed a filtering on automatically generated driver gene
lists per cohort. This filtering is intended to reduce artifacts from
the cohort-specific driver lists, due to e.g. errors in calling
algorithms, local hypermutation effects, undocumented filtering of
mutations.</p>
<p>We first created a collection of candidate driver genes by selecting
either: i) significant non-CGC genes (q-value &lt; 0.05) with at least two
significant bidders (methods rendering the genes as significant); ii)
significant CGC genes (either q-value &lt; 0.05 or CGC q-value &lt; 0.25) from
individual cohorts. All genes that did not fulfill these requirements
were discarded.</p>
<p>Additionally, candidate driver genes were further filtered using the
following criteria:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>We discarded non-expressed genes using TCGA expression data. For</dt><dd><p>tumor types directly mapping to cohorts from TCGA –including TCGA
cohorts– we removed non-expressed genes in that tumor type. We
used the following criterion for non-expressed genes: genes where
at least 80% of the samples showed a negative log2 RSEM. For those
tumor types which could not be mapped to TCGA cohorts this
filtering step was not done.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We also discarded genes highly tolerant to Single Nucleotide</dt><dd><p>Polymorphisms (SNP) across human populations. Such genes are more
susceptible to calling errors and should be taken cautiously. More
specifically, we downloaded transcript specific constraints from
gnomAD (release 2.1; 2018/02/14) and used the observed-to-expected
ratio score (oe) of missense (mys), synonymous (syn) and
loss-of-function (lof) variants to detect genes highly tolerant to
SNPs. Genes enriched in SNPs (oe_mys &gt; 1.5 or oe_lof &gt; 1.5 or
oe_syn &gt; 1.5) with a number of mutations per sample greater than 1
were discarded. Additionally, we discarded mutations overlapping
with germline variants (germline count &gt; 5) from a panel of
normals (PON) from Hartwig Medical Foundation
(<a class="reference external" href="https://nc.hartwigmedicalfoundation.nl/index.php/s/a8lgLsUrZI5gndd/download?path=%2FHMF-Pipeline-Resources&amp;files=SOMATIC_PON.vcf.gz">https://nc.hartwigmedicalfoundation.nl/index.php/s/a8lgLsUrZI5gndd/download?path=%2FHMF-Pipeline-Resources&amp;files=SOMATIC_PON.vcf.gz</a>).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We also discarded genes that are likely false positives according to</dt><dd><p>their known function from the literature. We convened that the
following genes are likely false positives: i) known long genes
such as TTN, OBSCN, RYR2, etc.; ii) olfactory receptors from HORDE
(<a class="reference external" href="http://bioportal.weizmann.ac.il/HORDE/">http://bioportal.weizmann.ac.il/HORDE/</a>; download date
2018/02/14); iii) genes not belonging to Tier1 CGC genes lacking
literature references according to CancerMine
(<a class="reference external" href="http://bionlp.bcgsc.ca/cancermine/">http://bionlp.bcgsc.ca/cancermine/</a>).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>We also removed non CGC genes with more than 3 mutations in one</dt><dd><p>sample. This abnormally high number of mutations in a sample may
be the result of either a local hypermutation process or cross
contamination from germline variants.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Finally we discarded genes whose mutations are likely the result of</dt><dd><p>local hypermutation activity. More specifically, some coding
regions might be the target of mutations associated to COSMIC
Signature 9 (<a class="reference external" href="https://cancer.sanger.ac.uk/cosmic/signatures">https://cancer.sanger.ac.uk/cosmic/signatures</a>)
which is associated to non-canonical AID activity in lymphoid
tumours. In those cancer types were Signature 9 is known to play a
significant mutagenic role (i.e., AML, Non-Hodgkin Lymphomas,
B-cell Lymphomas, CLL and Myelodysplastic syndromes) we discarded
genes where more than 50% of mutations in a cohort of patients
were associated with Signature 9.</p>
</dd>
</dl>
</li>
</ol>
<p>Candidate driver genes that were not discarded composed the catalog of
driver genes.</p>
<p>We then annotated the catalog of highly confident driver genes according
to their annotation level in CGC. Specifically, we created a three-level
annotation: i) the first level included driver genes with a reported
involvement in the source tumor type according to the CGC; ii) the
second group included CGC genes lacking reported association with the
tumor type; iii) the third group included genes that were not present in
CGC.</p>
<p>To match the tumor type of our analyzed cohorts and the
nomenclature/acronyms of cancer types reported in the CGC we manually
created a dictionary comprising all the names of tumor types from CGC
and cancer types defined in our study, according to the following rules:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>All the equivalent terms for a cancer type reported in the CGC using</dt><dd><p>the Somatic Tumor Type field (e.g. “breast”, “breast carcinoma”,
“breast cancer”), were mapped into the same tumor type.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>CGC terms with an unequivocal mapping into our cancer types were</dt><dd><p>automatically linked (e.g., “breast” with “BRCA”).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>CGC terms representing fine tuning classification of a more prevalent</dt><dd><p>cancer type that did not represent an independent cohort in our
study, were mapped to their closest parent tumor type in our study
(e.g., “malignant melanoma of soft parts” into “cutaneous
melanoma” or “alveolar soft part sarcoma” into “sarcoma”).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Adenomas were mapped to carcinomas of the same cell type</dt><dd><p>(e.g.,”hepatic adenoma” into “hepatic adenocarcinoma”, “salivary
gland adenoma” into “salivary gland adenocarcinoma”).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>CGC parent terms mapping to several tumor types from our study were</dt><dd><p>mapped to each of the potential child tumor types. For instance,
the term “non small cell lung cancer” was mapped to “LUAD” (lung
adenocarcinoma) and “LUSC” (lung squamous cell carcinoma).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Finally, CGC terms associated with benign lesions, with unspecified</dt><dd><p>tumor types (e.g., “other”, “other tumor types”, “other CNS”) or
with tumor types with missing parent in our study were left
unmatched.</p>
</dd>
</dl>
</li>
</ol>
<p>We computed the mode of action for highly confident driver genes. To do
so, we first performed a pan-cancer run of dNdScv across all TCGA
cohorts. We then applied the aforementioned algorithm (see Mode of
action section below for more details on how the algorithm determines
the role of driver genes according to their distribution of mutations in
a cohort of samples) to classify driver genes into the three possible
roles: Act (activating or oncogene), LoF (loss-of-function or tumor
suppressor) or Amb (ambiguous or non-defined). We then combined these
predictions with prior knowledge from the Cancer Genome Interpreter
[ref] according to the following rules: i) when the inferred mode of
action matched the prior knowledge, we used the consensus mode of
action; ii) when the gene was not included in the prior knowledge list,
we selected the inferred mode of action; iii) when the inferred mode of
action did not match the prior knowledge, we selected that of the prior
knowledge list.</p>
</div>
<div class="section" id="repository-of-mutational-features">
<h1>Repository of mutational features<a class="headerlink" href="#repository-of-mutational-features" title="Permalink to this headline">¶</a></h1>
<p>Linear clusters for each gene and cohort were identified by
OncodriveCLUSTL. We defined as significant those clusters in a driver
gene with a p-value lower than 0.05. The start and end of the clusters
were retrieved from the first and last mutated amino acid overlapping
the cluster, respectively.</p>
<p>Information about the positions involved in the 3D clusters defined by
HotMAPS were retrieved from the gene specific output of each cohort. We
defined as significant those amino acids in a driver gene with a q-value
lower than 0.05.</p>
<p>Pfam domains for each driver gene and cohort were identified by
smRegions. We defined as significant those domains in driver genes with
a q-value lower than 0.1 and with positive log ratio of
observed-to-simulated mutations (observed mutations / simulated
mutations &gt; 1). The first and last amino acid are defined from the start
and end of the Pfam domain, respectively.</p>
<p>The so-called excess of mutations for a given coding consequence-type
quantifies the proportion of observed mutations at this consequence-type
that are not explained by the neutral mutation rate. The excess is
inferred from the dN/dS estimate <span class="math notranslate nohighlight">\(\omega\)</span> as
<span class="math notranslate nohighlight">\((\omega - 1)\ /\ \omega\)</span>. We computed the excess for missense,
nonsense and splicing-affecting mutations.</p>
<p>We computed the gene-specific dN/dS estimates for nonsense and missense
mutations, denoted <span class="math notranslate nohighlight">\(\omega_{\text{non}}\)</span> and
<span class="math notranslate nohighlight">\(\omega_{\text{mis}}\)</span>. Then each gene induces a point in the
plane with coordinates
<span class="math notranslate nohighlight">\((\omega_{\text{non}},\ \omega_{\text{mis}})\)</span>. We deemed a gene
Act (resp. LoF) if its corresponding point sits above (resp. below) the
diagonal (<span class="math notranslate nohighlight">\(x = y\)</span>) up to an uncertainty threshold (XXX). Genes
within the uncertainty area as well as genes with
<span class="math notranslate nohighlight">\(\omega_{\text{non}} &lt; \ 1\)</span> and <span class="math notranslate nohighlight">\(\omega_{\text{mis}} &lt; \ 1\)</span>
were deemed “ambiguous”.</p>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, BBGLab.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="_static/language_data.js"></script>
      <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/latest.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  <script type="text/javascript">
      jQuery(function () {
          
          SphinxRtdTheme.Navigation.enableSticky();
          
      });
  </script> 

</body>
</html>